girder/api/v1/notification.py

```diff
@@ -71,8 +74,12 @@ def __init__(self):
         .produces('text/event-stream')
         .errorResponse()
         .errorResponse('You are not logged in.', 403)
+        .errorResponse('The notification stream is not enabled.', 503)
     )
     def stream(self, timeout, params):
+        if not Setting().get(SettingKey.ENABLE_NOTIFICATION_STREAM):
+            raise RestException('The notification stream is not enabled.', code=503)
```

Comment:
While "service unavailable" does seem to fit the semantics of our response, I think using status code 503 might not be a good idea here. I believe 5xx status codes typically indicate that something is *wrong* with the server, whereas in this case, nothing is wrong, it's just intentionally configured to be disabled. I wasn't going to object until I read the description [on wikipedia](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#5xx_Server_errors) of 503 in particular:

> The server is currently unavailable (because it is overloaded or down for maintenance). Generally, this is a temporary state.

As written, this isn't a temporary state, and it's not due to being overloaded or due to maintenance. We could add features in the future to temporarily throttle or disable the notifications due to load, and in that case I think 503 would be appropriate, but in this case I feel like status 410 may be our best bet.

Reply:
While a 500 indicates that something is probably wrong with the server, not all 5xx codes have the same semantics. From [RFC7231](https://tools.ietf.org/html/rfc7231#section-6.6):
> The 5xx (Server Error) class of status code indicates that the server is aware that it has erred or **is incapable of performing the requested method**.

For reference, here are definitions of the two codes in question:
* [410 Gone](https://httpstatusdogs.com/img/410.jpg)
> The 410 (Gone) status code indicates that access to the target resource is no longer available at the origin server and that this condition is likely to be permanent.  If the origin server does not know, or has no facility to determine, whether or not the condition is permanent, the status code 404 (Not Found) ought to be used instead.
>
> The 410 response is primarily intended to assist the task of web maintenance by notifying the recipient that the resource is intentionally unavailable and that the server owners desire that remote links to that resource be removed.  Such an event is common for limited-time, promotional services and for resources belonging to individuals no longer associated with the origin server's site. It is not necessary to mark all permanently unavailable resources as "gone" or to keep the mark for any length of time -- that is left to the discretion of the server owner.
>
> A 410 response is cacheable by default; i.e., unless otherwise indicated by the method definition or explicit cache controls.

* [503 Service Unavailable](https://httpstatusdogs.com/img/503.jpg)
> The 503 (Service Unavailable) status code indicates that the server is currently unable to handle the request due to a temporary overload or scheduled maintenance, which will likely be alleviated after some delay.  The server MAY send a Retry-After header field (Section 7.1.3) to suggest an appropriate amount of time for the client to wait before retrying the request.
>
> Note: The existence of the 503 status code does not imply that a server has to use it when becoming overloaded.  Some servers might simply refuse the connection.

As I understand it, the notification stream being disabled is:
* Caused by an action of the server, not the client
  * Accordingly, with a 503, the client was not 'at fault' for a perfectly reasonable, well-formed request, it's the server that's decided not to provide the resource.
  * To me, the implication of 410 Gone seems to be 'this resource isn't here, you shouldn't have asked for it, and you shouldn't ever ask for it again'.
* Not perminant
  * The decision to disable the notification stream is made by altering a runtime setting, and the functionality could be re-enabled at any time. Indeed, the actual policy of web clients is to attempt to reconnect to the notification stream every time the web app reloads.
  * While the server may send a `Retry-After` header with a 503, it is optional, and I see nothing wrong with a service being "temporarily" down, even if when or if it comes back up is uncertain, so long as the service *could* come back up (machines are very patient).
  * To restate my point from [our previous discussion about this](https://github.com/girder/girder/issues/2639):
    > In my understanding, "permanent" status codes (like 308 Permanent Redirect) are more for things where a policy is changed, and clients should ideally update themselves to never issue a request to the old route again.

Reply:
I still think this is more of a policy change than a temporary setting -- I think temporary conditions come with the expectation that they'll be "fixed" at some point. That said, I don't want to hold up this branch due to bikeshedding, so we can merge it as-is.